### ISSUE TITLE
Switch to Docker BuildX and BuildKit

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#9587]
As described at https://docs.docker.com/engine/release-notes/23.0/#new, `docker buildx` and BuildKit is the new default in Docker Engine 23.0+.

Based on the discussion starting at https://github.com/certbot/certbot/issues/8225#issuecomment-1429009879, there's some problems with our scripts when BuildKit is used so I'm proposing we stick to the legacy builder for now. As part of this change, I think we should also probably switch our `docker build` commands to `docker buildx build` and/or set `DOCKER_BUILDKIT=1` in our scripts so we they have consistent behavior across Docker Engine versions.

EDIT: After further investigation, things do not currently work with BuildKit and we should make this change because [the legacy Docker builder is deprecated](https://docs.docker.com/engine/deprecated/).

**Comment by bmw** [comment]
Actually, building with BuildKit may just Just Work™? See https://dev.azure.com/certbot/certbot/_build/results?buildId=6426&view=results.

Ideally we should play with this more and maybe proactively switch to using BuildKit, but since things seem to work both with and without it, I'm personally not going to worry about this for now.

**Comment by humanoid2050** [comment]
At this moment, no, simply switching to BuildKit with no additional changes probably will not work. This is due to the way BuildKit caches and accesses intermediate and base images. The separation of core certbot and the plugins into different Dockerfiles means that when you go to build plugins, BuildKit won't automagically find the certbot base image unless it has already been pushed to docker hub. However, this is a solvable problem with a couple different solutions.

**Comment by bmw** [comment]
Yeah? In the pipeline build I linked above (https://dev.azure.com/certbot/certbot/_build/results?buildId=6426&view=results), I configured things to use BuildKit, the images built successfully, and our simple smoke test passed.

There may be problems my test there didn't catch for whatever reason though (such as it being in the clean and ephemeral CI environment; repeated builds locally might fail).

**Comment by humanoid2050** [comment]
I think it succeeded by downloading the previous night's version of `docker.io/certbot/certbot:<arch>-nightly` from docker hub, not building off of the fresh image loaded in the docker image cache.

**Comment by bmw** [comment]
Ooo could be. It's not clear to me that that is happening from looking at the logs, but it's not clear that it's not happening either.

**Comment by humanoid2050** [comment]
I think if you were to run the pipeline with a tag that does not exist on docker hub (like `smoketest` or something), you would see an error arise.

**Comment by bmw** [comment]
> I think if you were to run the pipeline with a tag that does not exist on docker hub (like smoketest or something), you would see an error arise.

So I did that and my results were interesting! See https://dev.azure.com/certbot/certbot/_build/results?buildId=6450&view=results. The x86_64 builds passed but the ARM builds did not. I wonder why that is and if there's an easy tweak to get the ARM builds to work without much of a general overhaul of the system.

To respond to the rest of https://github.com/certbot/certbot/issues/8225#issuecomment-1441096554 though:

> In terms of real compute time, is the matrix strategy currently in use in the azure pipeline a parallel approach? As a corollary, how much parallelization does azure pipelines allow in practice? On my desktop with plenty of cores, use of buildkit means that all architectures get launched simultaneously, effectively resulting in a build time equivalent only to the slowest build.

This is a good question. We get 10 parallel jobs on Azure Pipelines. If we have jobs available, with the matrix approach, they'll run in parallel. It's not guaranteed that'll happen, but for instance, if you look at our nightly build last night, [this page says the build started at 8:30PM PST](https://dev.azure.com/certbot/certbot/_build/results?buildId=6449&view=results) and that's also when each of [the](https://dev.azure.com/certbot/certbot/_build/results?buildId=6449&view=logs&j=db9d590d-ed27-5cbb-1855-2594371ebd55) [Docker](https://dev.azure.com/certbot/certbot/_build/results?buildId=6449&view=logs&j=fdd3565a-f3c6-5154-eca9-9ae03666f7bd) [builds](https://dev.azure.com/certbot/certbot/_build/results?buildId=6449&view=logs&j=40dc3f3b-10cb-5f7f-1c5a-019f33f252fd) started. [Other builds](https://dev.azure.com/certbot/certbot/_build/results?buildId=6449&view=logs&j=3766a157-79e0-589e-e7a2-54bdfcd61823) started later.

[The Linux Azure Pipelines instances have two cores](https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml#hardware) though so in practice maybe it wouldn't actually be much slower?

> I'm honestly not sure when the platform flag was added to `docker pull`. A quick search didn't turn up any clues either. I'm actually not against having per-arch tags, I just wanted to ask the question to see if there were opportunities to simplify.

My personal preference for now I think is to leave them unless they become hard to create for some reason. If they do, we can look into the status of `docker pull --platform` on the oldest non-EOL'd version of CentOS/RHEL/Rocky, Debian, and Ubuntu to get a feel for the landscape there.

**Comment by humanoid2050** [comment]
> So I did that and my results were interesting! 

Very interesting indeed. The ARM failures are exactly what I expected to see. I have no idea why the native arch succeeded. 

Given the limitations of azure pipelines, there may not be a lot of actual time available to save. But still, good to know where the opportunities are. 

I have a full BuildKit branch at https://github.com/humanoid2050/certbot/tree/move-to-buildkit. I stripped things down to the max as a test to see what's possible. Things you might want to check out:

- All Docker operations have been collapsed into a single job in `packaging-jobs.yml`.
  - This includes calling `certbot plugins --prepare` on every image with every architecture as part of the Dockerfile
  - Pushing to docker hub occurs immediately after the images are built
  - BuildKit internally parallelizes the per-arch build operations
  - There are no artifacts stored and reloaded
- Getting the most out of BuildKit means using a docker-container driver. In docker 23+, the default builder is of this new type. Prior to that, the builder had to be created explicitly. See `build.sh` line 36.
- I have switched to using the `tonistiigi/binfmt` image to provide qemu components.
  - This image enables alternative architectures in the docker builder instance
  - Also allows non-native architecture images to run by directly invoking them
  - Removes the need to download qemu components and load them into the images

I should point out that the new qemu provider will not work its magic if `multiarch/qemu-user-static:register` has already registered interpreters. The `tonistiigi/binfm` image is the one suggested by https://docs.docker.com/build/building/multi-platform/.

It's also missing the per-arch tags, but I think it should be possible  to re-run the calls to `docker build` with single-arch platform specifications and the buildkit cache will quickly return the premade images for tagging. There might be a better way though.

**Comment by bmw** [comment]
Thanks so much for playing with this! The lines of code change being "82 additions and 266 deletions" is pretty appealing on top of all the other benefits here.

My high level feedback so far is:

1. I'd like us to be able to test building our Docker images in CI without actually deploying them. This was accomplished before by splitting things across `packaging-stage.yml` and `deploy-stage.yml` where those files are used as templates in other files under our top level `.azure-pipelines` directory. I don't care too much about the lower level details (except the ones I flagged in other comments here), but I think I'd probably like to keep this structure somehow.
2. I'd ideally like to continue testing the final image rather than testing things at build time. If that proves too cumbersome, we could potentially take this approach, but I'd want to make sure that any changes that Certbot command may make to directories like `/etc/letsencrypt` and `/var/log/letsencrypt` are not persisted in the underlying image.
3. My familiarity with the different builders and `tonistiigi/binfmt` is limited and I haven't yet done further digging (although I'm likely fine switching to both especially if that's what is recommended in Docker's documentation), but I'd like to make sure our script either a) cleans up nicely after itself so it can be run by devs locally or b) has some checks to prevent unfamiliar devs from running it. You may have already taken care of this.
4. While looking at things here, I noticed [`docker buildx build` accepts an `--output` flag](https://docs.docker.com/engine/reference/commandline/buildx_build/#output). Might it be possible to use this to keep our current high level build, test, and deploy structure? We may no longer need or want to split things by architecture though.

**Comment by humanoid2050** [comment]
Thanks for the input. I will play around with the layout. Seeing your use of azure pipelines for segmented testing definitely highlights the need for split build and deploy stages. There may be opportunities for separating steps without duplicating work by exporting the build cache and/or using appropriate `--output` arguments. I haven't fully explored all the output options, or more specifically what it takes to reload data from and an oci or multi-arch tar file.

**Comment by humanoid2050** [comment]
Sorry for the delay. Life got busy. Ive been mulling over the requirements for this project. I think the desire to have per-arch tags precludes using buildkit to parallelize the build process over the various architectures. As such, there will need to be a call to `docker build` for every arch+target combination. My question is whether you would want to expose this at the azure pipelines level, allowing the pipeline bandwidth to be maximized using a combination of jobs and matrix strategies, or keep everything in a shell script, possibly with something like GNU `parallel` to maximize throughput. Or we can do everything sequentially in loops and make it simple.

Either way, I think I can still provide some structural improvements with buildkit and the `tonistiigi/binfmt` image, including no longer having to manually download and place qemu pieces into the images. This saves about 2MB/image.

**Comment by bmw** [comment]
No worries. I don't think there's any rush here.

If us having per-arch tags significantly complicates things, I think we should drop them. Poking around Docker Hub, I didn't see any popular images with similar tags and instead everyone was just using a multi-arch manifest. Because of that, I think I'm fine with us taking that approach too.

If we do that though, we'll need to do something with the old tags. My initial idea is to delete them or upload them one last time including a printed warning when they are run saying that they will not be seeing updates in the future.

**Comment by humanoid2050** [comment]
I think it would be fine to leave old tags as they are, just re-release the primary <version> tag as a multiarch manifest. The latest tags disappear when new releases arrive anyway, and the arch-specific tags would remain for those pulling specific versions via script. Everyone else would either get the right image for their platform, or have the option to pull a specific platform. I think the only way someone would experience a breaking change is if they were using an arm platform and specifically expecting to get an amd64 image via the `certbot:<version>` tag. 
It might even be possible to automate the update process using `docker buildx imagetools create` to create manifests from the existing images on docker hub, no recompile necessary.

**Comment by humanoid2050** [comment]
I now know more about docker buildkit caching than I ever wanted to, but I think I have something over at https://github.com/humanoid2050/certbot/tree/build-with-buildkit.

Things you might find interesting
- There are separate build, test, and deploy scripts
- All docker operations are backed by buildkit
- To save work, docker artifacts are cached to the filesystem and reloaded on demand
- The full build process still takes about 23 minutes on my desktop
  - There might be ways to parallelized the building of plugins, but it sounds like the azure pipeline will be saturated regardless
- A lot of common functionality was moved into lib/common
- By removing qemu components, every image is about 2MB smaller (and less to download)

You can see the outputs at https://hub.docker.com/r/humanoid2050/certbot/tags.

The azure pipeline components still need to be updated.

Let me know what you think.

**Comment by bmw** [comment]
My concern with just leaving the old `latest-<arch>` tags as they are is that they'll never receive updates which users of those tags may not know about. I'm personally leaning towards pushing a final update that does something like what we did at https://github.com/certbot/certbot/pull/7187 which I think should be easy enough. Alternatively we could just delete them but that's maybe a little aggressive. I'm not sure what the best/common practices are here to the degree they exist :shrug:

At a high level your code looks pretty great to me. I hope you found the deep dive into docker buildkit somewhat fun, useful, or interesting. Regardless, thanks so much for doing it. It's a big help here.

The only thing I saw that I at least initially wished was a little different is I'm not sure I like how `test.sh` builds the Docker images if they don't already exist on your system. I worry this could result in us accidently building and testing different images in CI/CD than the ones we deploy to Docker Hub in the future.

If possible, I'd like to change that. We could keep the current behavior conditional on a CLI argument or environment variable if that's necessary or useful for some reason. Alternatively, if telling Docker to load/run images from the build cache without building them is tricky, maybe we could add some extra code that checks if the images already exists and errors out if they do not. I also think this all may become easier once exporting/importing the images through Azure Pipelines is figured out.

**Comment by wgreenberg** [comment]
re: what to do with older images, i think deleting them is fine so long as we announce it ahead of time in an easily-googleable way. adding a warning message can only help, although i suspect many people won't even see it until the image goes missing and things break.

**Comment by humanoid2050** [comment]
> concern with just leaving the old latest-<arch> tags

I forgot about that tag. However, it should probably be pointed out that the tag takes up virtually no overhead, and the blob referenced is on the hub anyhow. It is nice to have fewer things to track though, and as far as I can tell, making these individual tags would require pushing each tag individually. I think the image transfer protocols should automatically detect that the tag references an existing blob and not actually push layers. 

> I'm not sure I like how test.sh builds the Docker images

I also dislike that "feature". The `deploy.sh` script can also do the same thing.  Cards on the table, I'm not really a fan of the cache-based approach to avoiding rebuilding of images. Any changes, even changing the permissions on a script, forces a rebuild of the entire stack of images. However, I'm not sure if I like any of the other solutions any better.

- Alternative 1 is to load complete images to the local docker cache. This has to happen for each arch separately, as the docker cache does not support multiarch manifests. Then the test phase could deterministically reference each image. When it comes time to deploy the scripts, each image would also have to be uploaded separately, and the manifest generated at the end via `docker buildx imagetools create`. This is pretty much what currently exists as far as pipeline layout goes.
- Alternative 2 is to save images from the build stage to a temporary docker registry accessible by the azure pipeline. This would require the test stage to `docker pull --platform <arch>` each image for testing. I also think the deploy stage would have the same overhead as Alternative 1 of pulling and pushing every layer individually, then building the manifest separately.
- Alternative 3 is to directly load images to the docker registry under a tag like "testing". There is a risk of uploading a broken image, but hopefully the tag discourages use. The test stage would then `docker pull --platform <arch>` each image that was just uploaded. If all tests pass, I think the `imagetools` command could be use to rename the remote manifests with nightly or a version stamp in the deploy stage. This pattern is less amenable to continuing with `latest-<arch>` tags, as it would require parsing the manifest contents to get image digests for each architecture as it exists on the registry (or track and accumulate that information during the test phase when all images get pulled separately).

Given these alternatives, where does the automatic rebuild "feature" of `test.sh` and `deploy.sh` land?

**Comment by bmw** [comment]
If keeping the `latest-<arch>` tags is relatively easy, I think we should do so. If it's tricky or adds a fair bit of complexity, I think we should gently deprecate/remove them. I personally suggest we ignore this problem for now, develop a solution for switching to buildx without worrying about them, and then decide what to do based on how hard it'd be to keep them.

I think my personal preference is (1) or (2) largely depending on which is cleaner/easier to implement. 
The thing I don't like about (3) is we'd have to more widely expose our Docker Hub credentials in CI. Given we currently have something like (1), I suspect that'd be easier than (2), but if it's not, I'm certainly open to something like that.

I'm surprised by how tricky (1) seems to be though. After running your current commands, the data for both the single images and the multiarch manifest exist somewhere on the system, but it seems like there isn't a good way to move that data around without publishing it to a registry? Do none of the [`docker buildx build --output` formats](https://docs.docker.com/engine/reference/commandline/buildx_build/#output) and/or commands like `docker save` help? It's also worth flagging that each "job" in Azure Pipelines executes in a clean environment so to avoid rebuilding with the current approach, I think we'd have to do something to copy the Docker cache between jobs :weary:

**Comment by humanoid2050** [comment]
Seems like there isn't a perfect way forward. So how about this. I will set the build process to invoke buildkit using `docker buildx build`. I will first call it for all platforms in parallel, leaving all images in the build cache.  Then I will do per-arch calls to `docker buildx build --load` to pull each image into the docker cache with an appropriate tag. The test stage will exclusively reference the docker cache for each image to be tested. The deploy stage will also exclusively reference the docker cache as well, and manually create the multiarch manifest at the end using `docker buildx imagetools create`. 

That will check all the boxes, even if it does leave some room for improvement with parallelization later. It will also allow us to remove the static QEMU libraries, saving about 2M per image.

**Comment by bmw** [comment]
A couple clarifying questions:

1. How do you plan to move the Docker cache between jobs given what I said in my last message about each "job" in Azure Pipelines executing in a clean environment? By that I mean the Docker cache will not automatically persist between the build, test, and deploy steps.
2. Will `test.sh` and `deploy.sh` still rebuild the image if they can't find it? I'd really like to avoid this and I'm not sure I understand what you mean when you say that these steps will exclusively reference the Docker cache.

You may have a better idea in mind than me depending on the answer to those questions, but a possible solution I see here is to use `docker buildx build` in `build.sh` to build the images with a value for `--output` that we can easily copy the result and then use `docker load` in the test and deploy stages to load the result without using `docker buildx build` in those steps to avoid rebuilding. I think this ends up being very close to what we currently have in this repo except we're switching `build.sh` to use `docker buildx build` and maybe doing less splitting by architecture. It should also hopefully make keeping the `latest-<arch>` tags pretty easy.

I also noticed that Docker documents the `--load` argument to `docker buildx build` as just "Shorthand for `--output=type=docker`" which I found interesting.

**Comment by humanoid2050** [comment]
Sorry for the delay. Life got busy.

To answer your questions. 

> How do you plan to move the Docker cache

I should be clear that there are two caches in play: the build cache and the docker image cache. I was planning to migrate the build cache by directing buildkit to store the cache on the file system with the `--cache-to=dir` option, and then taring/untaring the directory across pipeline jobs. This is similar to the way the image cache is currently migrated, but it doesn't have to happen per architecture.

> Will test.sh and deploy.sh still rebuild the image

No. The plan was to switch away from migrating the build cache, and continue migrating the docker image cache between pipeline jobs. 

Basically, your suggestion of using `--load` to populate the docker image cache and then continue with `docker save`/`docker load` to migrate images across pipeline jobs is exactly where I was going. I just explained it poorly. 

THAT BEING SAID. I found something new: `docker buildx bake`. It parallelizes all of the named targets in a group, meaning no more for-loop to generate all of the plugin images! I'm testing it now. Looks like it brings the build stage all the way down to about 15 minutes for all architectures on my computer. I'm getting an error in the test stage though. Not sure if it's related. I'm also tracking down some intermittent conditions under which images get rebuilt at the image loading stage instead of being pulled from cache.

My current code is at https://github.com/humanoid2050/certbot/tree/build-with-buildkit

**Comment by bmw** [comment]
Sticking with `docker save` and `docker load` sounds good to me as does the improvements of `docker buildx bake`. Nice find there.

If you have any questions or could use a 2nd set of eyes on something in particular, please don't hesitate to let me know.

**Comment by humanoid2050** [comment]
Actually, if you have some spare CPU cycles, could you checkout that branch I linked and run run

`docker rmi $(docker images --filter dangling=true -q) || ./build.sh test all && ./test.sh test all`

repeatedly?

I'm seeing a random error in the arm64 dnsimple image where it tries to reference the dns-google plugin. I don't know the internals of the plugins well enough to find where things are getting crossed. I'm also trying to get a sense of how often the cache gets ignored and layers get rebuilt during the `--load` phase of build. The `docker rmi` just removes untagged docker images. Otherwise you are going to get a lot of them taking up space as the tags get migrated to each successive build.

**Comment by bmw** [comment]
It took some finagling and false starts for me to be able to run it repeatedly (e.g. I think you need to replace `docker builder rm certbot_builder` with `docker buildx rm certbot_builder`), but once I did, it failed for me on the first run with:
```
+ docker run --rm certbot/dns-dnsimple:arm64v8-test plugins --prepare
WARNING: The requested image's platform (linux/arm64) does not match the detected host platform (linux/amd64) and no specific platform was requested
An unexpected error occurred:
pkg_resources.DistributionNotFound: The 'dns-lexicon>=3.2.1' distribution was not found and is required by the application
Ask for help or search for solutions at https://community.letsencrypt.org. See the logfile /tmp/certbot-log-_nxual0b/log or re-run Certbot with -v for more details.
```
Is that what you're seeing? dnsimple and all of our plugin images more generally may have similar intermediate build steps, but in terms of real dependencies, they should all be completely independent of each other and only dependent on the core Certbot image. For what it's worth, the core Certbot image shouldn't (and probably doesn't) have the `dns-lexicon` library the error message is complaining about, however, that should have been installed as part of building the dnsimple plugin image.

**Comment by bmw** [comment]
Hmm...so the potential of `docker buildx bake` seems pretty cool, but since we're having problems with it, what do you think about just sticking with `docker buildx build` for now?

**Comment by humanoid2050** [comment]
On a personal level, it feels like defeat. Haha. But fortunately my personal feelings don't dictate policy for this project! So I will dial back the `docker buildx bake` to a bash for-loop and leave a note that there may be future improvements.

**Comment by bmw** [comment]
I mean you're welcome to keep playing with it if you want! Switching to `docker buildx build` is a smaller change that would resolve the original purpose of this issue though.

If you do decide you want to keep playing with `bake`, one thing I haven't seen which may be useful is the complete build log for all images that results in problems like this. I think investigating that could be interesting.

**Comment by bmw** [comment]
> I will probably make a comment with the git hash so that someone can refer to the existing bake stuff in the future when maybe it's more stable.

This sounds good! If you point me to the commit you have in mind, I can make a branch containing it here on the Certbot repo that the comment can point to.

**Comment by bmw** [comment]
This was fixed in #9628.
